sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ComplexTypes.scala

```diff
@@ -39,19 +40,14 @@ object SimplifyExtractValueOps extends Rule[LogicalPlan] {
       // Remove redundant field extraction.
       case GetStructField(createNamedStruct: CreateNamedStruct, ordinal, _) =>
         createNamedStruct.valExprs(ordinal)
-      case GetStructField(w @ WithFields(struct, names, valExprs), ordinal, maybeName) =>
-        val name = w.dataType(ordinal).name
-        val matches = names.zip(valExprs).filter(_._1 == name)
-        if (matches.nonEmpty) {
-          // return last matching element as that is the final value for the field being extracted.
-          // For example, if a user submits a query like this:
-          // `$"struct_col".withField("b", lit(1)).withField("b", lit(2)).getField("b")`
-          // we want to return `lit(2)` (and not `lit(1)`).
-          val expr = matches.last._2
-          If(IsNull(struct), Literal(null, expr.dataType), expr)
-        } else {
-          GetStructField(struct, ordinal, maybeName)
-        }
+    case GetStructField(updateFields: UpdateFields, ordinal, _) =>
+      val structExpr = updateFields.structExpr
+      updateFields.newExprs(ordinal) match {
+        // if the struct itself is null, then any value extracted from it (expr) will be null
+        // so we don't need to wrap expr in If(IsNull(struct), Literal(null, expr.dataType), expr)
+        case expr: GetStructField if expr.child.semanticEquals(structExpr) => expr
+        case expr => If(IsNull(ultimateStruct(structExpr)), Literal(null, expr.dataType), expr)
```

Comment:
shall we apply this rule after `UpdateFields` are all merged? then we don't need to do `ultimateStruct`.

Reply:
IIUC you mean put `CombineUpdateFields` in a separate batch that runs before the `SimplifyExtractValueOps` batch and remove the `ultimateStruct` method?

If so, we'll miss out on some optimizations because after `SimplifyExtractValueOps` runs we might again end up with code containing `UpdateFields(UpdateFields(_, _))`. The `simplify add multiple nested fields to struct` test in `complexTypesSuite` is a good example of a test that will fail in this scenario.

Reply:
I mean something like
```
case GetStructField(updateFields: UpdateFields, ordinal, _) if !updateFields.structExpr.isInstanceOf[UpdateFields] =>
  ...
```
As we know that there must be a rule to merge adjacent `UpdateFields` and this rule can wait for it to happen. They are still in the same batch.

Reply:
oh I see, passes my tests so I have made the change.